sockeye/beam_search.py

```diff
@@ -30,6 +30,10 @@
 
 class _Inference(ABC):
 
+    @abstractmethod
+    def state_structure(self):
```

Comment:
It seems that we now got rid of the nested state lists in SortStates. This also means that we need to return a flat list of states in the case of ensembles. How about having `state_structure` always return a string, and `encode_and_initialize` and `decode_step` always a flat list of states? This would require changes in lines 93 and 108 (`+=` instead of `append`).

Reply:
Actually, now that I'm considering it more, it may be better to revert back to using treating nested state lists in SortStates. I have a quick question about that though, in that case, are the model states guaranteed to be either a list of arrays (single model), or a list of lists of arrays (ensemble)? Are there any cases of lists of lists of lists of arrays, etc?

Reply:
Yes, there is no arbitrary depth of nesting, just 2 levels in case of ensembling.